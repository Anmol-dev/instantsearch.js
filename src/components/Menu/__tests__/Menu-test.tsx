/** @jsx h */

import { h } from 'preact';
import { render, fireEvent } from '@testing-library/preact';
import Menu from '../Menu';

const cssClasses = {
  root: 'root',
  noRefinementRoot: 'noRefinementRoot',
  list: 'list',
  item: 'item',
  selectedItem: 'selectedItem',
  link: 'link',
  label: 'label',
  count: 'count',
  showMore: 'showMore',
  disabledShowMore: 'disabledShowMore',
};

const templates = {
  item:
    '<a class="{{cssClasses.link}}" href="{{url}}">' +
    '<span class="{{cssClasses.label}}">{{label}}</span>' +
    '<span class="{{cssClasses.count}}">{{count}}</span>' +
    '</a>',
  showMoreText: `
    {{#isShowingMore}}
      Show less
    {{/isShowingMore}}
    {{^isShowingMore}}
      Show more
    {{/isShowingMore}}
  `,
};

const defaultProps = {
  canToggleShowMore: false,
  createURL: value => `#${value}`,
  cssClasses,
  isShowingMore: false,
  items: [],
  showMore: false,
  templateProps: {
    templates,
  },
  toggleRefinement: () => {},
  toggleShowMore: () => {},
};

describe('Menu', () => {
  test('renders correct DOM without items', () => {
    const props = {
      ...defaultProps,
    };

    const { container } = render(<Menu {...props} />);
    const list = container.querySelector('.list');

    expect(list).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders correct DOM with items', () => {
    const items = [
      {
        label: 'Appliances',
        value: 'Appliances',
        count: 4306,
        isRefined: true,
      },
      {
        label: 'Computers & Tablets',
        value: 'Computers & Tablets',
        count: 3563,
        isRefined: false,
      },
      {
        label: 'Cell Phones',
        value: 'Cell Phones',
        count: 3291,
        isRefined: false,
      },
      {
        label: 'Cell Phone Accessories',
        value: 'Cell Phone Accessories',
        count: 2836,
        isRefined: false,
      },
      {
        label: 'Audio',
        value: 'Audio',
        count: 1570,
        isRefined: false,
      },
      {
        label: 'Small Kitchen Appliances',
        value: 'Small Kitchen Appliances',
        count: 1510,
        isRefined: false,
      },
      {
        label: 'Cameras & Camcorders',
        value: 'Cameras & Camcorders',
        count: 1369,
        isRefined: false,
      },
      {
        label: 'Car Electronics & GPS',
        value: 'Car Electronics & GPS',
        count: 1208,
        isRefined: false,
      },
      {
        label: 'TV & Home Theater',
        value: 'TV & Home Theater',
        count: 1201,
        isRefined: false,
      },
      {
        label: 'Cell Phone Cases & Clips',
        value: 'Cell Phone Cases & Clips',
        count: 1195,
        isRefined: false,
      },
    ];
    const props = {
      ...defaultProps,
      items,
    };

    const { container } = render(<Menu {...props} />);
    const list = container.querySelector('.list');
    const listItems = container.querySelectorAll('.item');

    expect(list).toBeInTheDocument();
    expect(listItems).toHaveLength(items.length);
    expect(container).toMatchSnapshot();
  });

  test('calls templates with correct data', () => {
    const itemTemplate = jest.fn(() => '');
    const props = {
      ...defaultProps,
      items: [
        {
          label: 'Appliances',
          value: 'Appliances',
          count: 4306,
          isRefined: false,
        },
      ],
      templateProps: {
        templates: {
          item: itemTemplate,
        },
      },
    };

    render(<Menu {...props} />);

    expect(itemTemplate).toHaveBeenCalledTimes(1);
    expect(itemTemplate).toHaveBeenCalledWith({
      count: 4306,
      cssClasses,
      isRefined: false,
      label: 'Appliances',
      url: '#Appliances',
      value: 'Appliances',
    });
  });

  test('triggers refinement on click', () => {
    const toggleRefinement = jest.fn();
    const props = {
      ...defaultProps,
      items: [
        {
          label: 'Appliances',
          value: 'Appliances',
          count: 4306,
          isRefined: false,
        },
      ],
      toggleRefinement,
    };

    const { container } = render(<Menu {...props} />);
    const item = container.querySelector('.item')!;

    fireEvent.click(item);

    expect(toggleRefinement).toHaveBeenCalledTimes(1);
    expect(toggleRefinement).toHaveBeenCalledWith('Appliances');
  });

  test('renders with showMore and isShowingMore to false', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
      canToggleShowMore: true,
    };

    const { container } = render(<Menu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show more');
    expect(container).toMatchSnapshot();
  });

  test('renders with showMore and isShowingMore to true', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: true,
      canToggleShowMore: true,
    };

    const { container } = render(<Menu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show less');
    expect(container).toMatchSnapshot();
  });

  test('renders with showMore and canToggleShowMore to false', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
    };

    const { container } = render(<Menu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show more');
    expect(showMore).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  test('calls toggleShowMore on show more button click', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
      canToggleShowMore: true,
      toggleShowMore: jest.fn(),
    };

    const { container } = render(<Menu {...props} />);
    const showMore = container.querySelector('.showMore')!;

    fireEvent.click(showMore);

    expect(props.toggleShowMore).toHaveBeenCalledTimes(1);
  });
});
