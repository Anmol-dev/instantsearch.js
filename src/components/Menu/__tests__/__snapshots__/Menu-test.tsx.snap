// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correct DOM with items 1`] = `
<div>
  <div
    class="root"
  >
    <ul
      class="list"
    >
      <li
        class="item selectedItem"
      >
        <div>
          <a
            class="link"
            href="#Appliances"
          >
            <span
              class="label"
            >
              Appliances
            </span>
            <span
              class="count"
            >
              4306
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Computers & Tablets"
          >
            <span
              class="label"
            >
              Computers & Tablets
            </span>
            <span
              class="count"
            >
              3563
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cell Phones"
          >
            <span
              class="label"
            >
              Cell Phones
            </span>
            <span
              class="count"
            >
              3291
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cell Phone Accessories"
          >
            <span
              class="label"
            >
              Cell Phone Accessories
            </span>
            <span
              class="count"
            >
              2836
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Audio"
          >
            <span
              class="label"
            >
              Audio
            </span>
            <span
              class="count"
            >
              1570
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Small Kitchen Appliances"
          >
            <span
              class="label"
            >
              Small Kitchen Appliances
            </span>
            <span
              class="count"
            >
              1510
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cameras & Camcorders"
          >
            <span
              class="label"
            >
              Cameras & Camcorders
            </span>
            <span
              class="count"
            >
              1369
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Car Electronics & GPS"
          >
            <span
              class="label"
            >
              Car Electronics & GPS
            </span>
            <span
              class="count"
            >
              1208
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#TV & Home Theater"
          >
            <span
              class="label"
            >
              TV & Home Theater
            </span>
            <span
              class="count"
            >
              1201
            </span>
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cell Phone Cases & Clips"
          >
            <span
              class="label"
            >
              Cell Phone Cases & Clips
            </span>
            <span
              class="count"
            >
              1195
            </span>
          </a>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Menu renders correct DOM without items 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  />
</div>
`;

exports[`Menu renders with showMore and canToggleShowMore to false 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore disabledShowMore"
      disabled=""
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`Menu renders with showMore and isShowingMore to false 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore"
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`Menu renders with showMore and isShowingMore to true 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore"
    >
      Show less
    </button>
  </div>
</div>
`;
